File: openry/orchestrator/router.py
"""Phase 3a: Conditional routing engine with when / when_any short-circuit evaluation.

Evaluates validation_routing entries from a sub_step YAML config, determines
the routing target based on short-circuit AND/OR logic.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from openry.db import get_task_state

from .validator import ValidationContext, ValidationResult, validate


# ── Data structures ────────────────────────────────────────────


@dataclass
class RoutingResult:
    """Result of evaluating the entire validation_routing block.

    Attributes:
        action: What to do next.
            - "route": Proceed to the specified target.
            - "fallthrough": No routing decision made (delegate to Phase 2).
        target: Routing target if action is "route".
            - "done": Big step complete (success).
            - "abort": Big step failed.
            - "retry_current": Retry current sub_step.
            - "continue": Internal signal to proceed to next entry.
            - Any sub_step_id string.
        message: Human-readable description of the routing decision.
    """

    action: str  # "route" | "fallthrough"
    target: str = ""
    message: str = ""

    @classmethod
    def route_to(cls, target: str, message: str = "") -> RoutingResult:
        return cls(action="route", target=target, message=message)

    @classmethod
    def fallthrough(cls, message: str = "") -> RoutingResult:
        return cls(action="fallthrough", message=message)


# ── Valid routing targets ──────────────────────────────────────

VALID_TARGETS = frozenset({"done", "abort", "retry_current", "continue"})


def _is_valid_target(target: str) -> bool:
    """Check if a routing target is one of the well-known keywords or a sub_step ID."""
    if target in VALID_TARGETS:
        return True
    # Non-keyword targets are assumed to be sub_step IDs
    return bool(target and target not in ("",))


# ── Main entry point ───────────────────────────────────────────


def evaluate_routing(run_id: str, step_config: dict[str, Any]) -> RoutingResult:
    """Evaluate all validation_routing entries with short-circuit logic.

    Processing order:
      1. For each entry (when or when_any), evaluate conditions.
      2. If ANY entry fails (on_mismatch), route immediately (short-circuit).
      3. If ALL entries pass, use the global on_success.

    Validation errors (exceptions during evaluation) cause the entry
    to be skipped; if all entries error out, fall through to Phase 2.

    Args:
        run_id: The task run_id being validated.
        step_config: The sub_step YAML configuration dict.

    Returns:
        RoutingResult with action and target.
    """
    ctx = _build_context(run_id)

    entries = step_config.get("validation_routing", [])
    if not entries:
        return RoutingResult.fallthrough("no validation_routing entries")

    error_count = 0

    for idx, entry in enumerate(entries):
        has_when = "when" in entry
        has_when_any = "when_any" in entry

        if not has_when and not has_when_any:
            error_count += 1
            continue

        if has_when_any:
            result = _evaluate_when_any(ctx, entry)
        else:
            result = _evaluate_when(ctx, entry)

        if result.action == "error":
            error_count += 1
            continue

        if result.action == "route":
            if result.target == "continue":
                continue
            return result

        return result

    if error_count == len(entries):
        return RoutingResult.fallthrough("all validation_routing entries errored")

    on_success = step_config.get("on_success", "done")
    return RoutingResult.route_to(on_success, "all validation_routing entries passed")


# ── Entry evaluators ───────────────────────────────────────────


def _evaluate_when(
    ctx: ValidationContext, entry: dict[str, Any],
) -> RoutingResult:
    """Evaluate a single 'when' entry (one condition)."""
    condition = entry["when"]
    result = validate(ctx, condition)

    if result.passed:
        on_match = entry.get("on_match", "continue")
        if on_match == "continue":
            return RoutingResult.route_to("continue", "condition passed, continue")
        if _is_valid_target(on_match):
            return RoutingResult.route_to(on_match, result.message or "condition passed")
        return RoutingResult.fallthrough(f"unknown on_match target: {on_match}")
    else:
        on_mismatch = entry.get("on_mismatch", "abort")
        msg = entry.get("on_mismatch_message", result.message)
        if _is_valid_target(on_mismatch):
            return RoutingResult.route_to(on_mismatch, msg)
        return RoutingResult.fallthrough(f"unknown on_mismatch target: {on_mismatch}")


def _evaluate_when_any(
    ctx: ValidationContext, entry: dict[str, Any],
) -> RoutingResult:
    """Evaluate a 'when_any' entry (OR group: any sub-condition passes → pass)."""
    conditions = entry.get("when_any", [])
    if not conditions:
        # Empty when_any → treated as failed
        on_mismatch = entry.get("on_mismatch", "abort")
        msg = entry.get("on_mismatch_message", "empty when_any group")
        if _is_valid_target(on_mismatch):
            return RoutingResult.route_to(on_mismatch, msg)
        return RoutingResult.fallthrough(f"unknown on_mismatch target: {on_mismatch}")

    errors: list[str] = []
    any_passed = False

    for condition in conditions:
        try:
            result = validate(ctx, condition)
        except Exception as exc:
            errors.append(str(exc))
            continue

        if result.passed:
            any_passed = True
            break  # OR: first pass wins
        else:
            errors.append(result.message)

    if any_passed:
        on_match = entry.get("on_match", "continue")
        if on_match == "continue":
            return RoutingResult.route_to("continue", "when_any: condition passed")
        if _is_valid_target(on_match):
            return RoutingResult.route_to(on_match, "when_any: condition passed")
        return RoutingResult.fallthrough(f"unknown on_match target: {on_match}")
    else:
        on_mismatch = entry.get("on_mismatch", "abort")
        msg = entry.get(
            "on_mismatch_message",
            f"when_any: no conditions matched ({'; '.join(errors)})",
        )
        if _is_valid_target(on_mismatch):
            return RoutingResult.route_to(on_mismatch, msg)
        return RoutingResult.fallthrough(f"unknown on_mismatch target: {on_mismatch}")


# ── Helpers ────────────────────────────────────────────────────


def _build_context(run_id: str) -> ValidationContext:
    """Build a ValidationContext from a run_id by reading DB state."""
    state = get_task_state(run_id)
    if not state:
        return ValidationContext(run_id=run_id, payload={})

    try:
        payload = json.loads(state.get("payload", "{}"))
    except (json.JSONDecodeError, TypeError):
        payload = {}

    return ValidationContext(
        run_id=run_id,
        payload=payload,
        step_config={},  # step_config passed separately at call site
    )

File: orchestrator-plugin/src/orchestrator/router.ts
/**
 * Phase 3a: 条件路由引擎 — when / when_any 短路求值。
 *
 * 与 Python openry/orchestrator/router.py 逻辑完全一致。
 */
import Database from "better-sqlite3";
import type { ValidationRoutingEntry } from "./yaml-loader.js";
import { validateCondition, type ValidationRule } from "./validator.js";

// ── Types ──────────────────────────────────────────────────────

export type RoutingResult = {
  action: "route" | "fallthrough";
  target: string;
  message: string;
};

// ── Valid targets ──────────────────────────────────────────────

const VALID_TARGETS = new Set(["done", "abort", "retry_current", "continue"]);

function isValidTarget(t: string): boolean {
  if (VALID_TARGETS.has(t)) return true;
  return t !== "" && t != null;
}

// ── Build context ──────────────────────────────────────────────

function loadPayload(db: Database.Database, runId: string): Record<string, unknown> {
  const row = db
    .prepare("SELECT payload FROM task_state WHERE run_id = ?")
    .get(runId) as { payload: string } | undefined;
  if (!row) return {};
  try {
    return JSON.parse(row.payload);
  } catch {
    return {};
  }
}

// ── Main entry ─────────────────────────────────────────────────

export async function evaluateRouting(
  db: Database.Database,
  runId: string,
  stepConfig: { on_success?: string; on_failure?: string; validation_routing?: ValidationRoutingEntry[] },
): Promise<RoutingResult> {
  const entries = stepConfig.validation_routing ?? [];
  if (entries.length === 0) {
    return { action: "fallthrough", target: "", message: "no validation_routing entries" };
  }

  const payload = loadPayload(db, runId);
  let errorCount = 0;

  for (const entry of entries) {
    const hasWhen = "when" in entry && entry.when;
    const hasWhenAny = "when_any" in entry && entry.when_any;

    if (!hasWhen && !hasWhenAny) {
      errorCount += 1;
      continue;
    }

    let result: RoutingResult;
    try {
      if (hasWhenAny) {
        result = await evaluateWhenAny(payload, entry);
      } else {
        result = await evaluateWhen(payload, entry);
      }
    } catch {
      errorCount += 1;
      continue;
    }

    if (result.action === "route") {
      if (result.target === "continue") continue; // internal pass signal
      return result; // real routing decision (mismatch or match → direct route)
    }
    // fallthrough → let Phase 2 handle
    return result;
  }

  if (errorCount === entries.length) {
    return { action: "fallthrough", target: "", message: "all validation_routing entries errored" };
  }

  const onSuccess = stepConfig.on_success ?? "done";
  return { action: "route", target: onSuccess, message: "all validation_routing entries passed" };
}

// ── Entry evaluators ───────────────────────────────────────────

function evaluateWhen(
  payload: Record<string, unknown>,
  entry: ValidationRoutingEntry,
): Promise<RoutingResult> {
  const condition = entry.when as ValidationRule;
  return validateCondition(payload, condition).then((result) => {
    if (result.passed) {
      const onMatch = entry.on_match ?? "continue";
      if (onMatch === "continue") {
        return { action: "route" as const, target: "continue", message: "condition passed, continue" };
      }
      if (isValidTarget(onMatch)) {
        return { action: "route" as const, target: onMatch, message: result.message || "condition passed" };
      }
      return { action: "fallthrough" as const, target: "", message: `unknown on_match target: ${onMatch}` };
    }

    const onMismatch = entry.on_mismatch ?? "abort";
    const msg = entry.on_mismatch_message ?? result.message;
    if (isValidTarget(onMismatch)) {
      return { action: "route" as const, target: onMismatch, message: msg };
    }
    return { action: "fallthrough" as const, target: "", message: `unknown on_mismatch target: ${onMismatch}` };
  });
}

async function evaluateWhenAny(
  payload: Record<string, unknown>,
  entry: ValidationRoutingEntry,
): Promise<RoutingResult> {
  const conditions = (entry.when_any ?? []) as ValidationRule[];
  if (conditions.length === 0) {
    const onMismatch = entry.on_mismatch ?? "abort";
    const msg = entry.on_mismatch_message ?? "empty when_any group";
    if (isValidTarget(onMismatch)) {
      return { action: "route", target: onMismatch, message: msg };
    }
    return { action: "fallthrough", target: "", message: `unknown on_mismatch target: ${onMismatch}` };
  }

  const errors: string[] = [];
  for (const condition of conditions) {
    try {
      const result = await validateCondition(payload, condition);
      if (result.passed) {
        // OR: first pass wins
        const onMatch = entry.on_match ?? "continue";
        if (onMatch === "continue") {
          return { action: "route", target: "continue", message: "when_any: condition passed" };
        }
        if (isValidTarget(onMatch)) {
          return { action: "route", target: onMatch, message: "when_any: condition passed" };
        }
        return { action: "fallthrough", target: "", message: `unknown on_match target: ${onMatch}` };
      }
      errors.push(result.message);
    } catch (err) {
      errors.push(String(err));
    }
  }

  // All conditions failed
  const onMismatch = entry.on_mismatch ?? "abort";
  const msg = entry.on_mismatch_message ?? `when_any: no conditions matched (${errors.join("; ")})`;
  if (isValidTarget(onMismatch)) {
    return { action: "route", target: onMismatch, message: msg };
  }
  return { action: "fallthrough", target: "", message: `unknown on_mismatch target: ${onMismatch}` };
}
